### PR TITLE
feat(connector): Add topic to mqtt additional columns

### PR DESCRIPTION
**Pull request by Boudewijn26**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Make the MQTT topic available as an additional column
- Revert the change made in https://github.com/risingwavelabs/risingwave/pull/18961 to restore MQTT functionality

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

* It's now possible to include the topic on which MQTT messages have been received by using `INCLUDE topic` in your `SELECT`

### CODE REVIEW

File: src/connector/src/source/mqtt/enumerator/mod.rs [hunk 1]
```diff
@@ -117,10 +119,18 @@ impl SplitEnumerator for MqttSplitEnumerator {
...
+                if self.connected.load(std::sync::atomic::Ordering::Relaxed) {
+                    break;
+                }
+
+                if start.elapsed().as_secs() > 10 {
+                    bail!("Failed to connect to mqtt broker");
+                }
+
+                tokio::time::sleep(std::time::Duration::from_millis(500)).await;
+            }
```
Comment (tabVersion) [inline src/connector/src/source/mqtt/enumerator/mod.rs:133]: regarding the comments https://github.com/risingwavelabs/risingwave/pull/18961#issuecomment-2422455694, I still have some concern about the wait time. Would you be able to move the logic in the tokio thread into this func?

File: src/connector/src/parser/additional_columns.rs [hunk 2]
```diff
@@ -87,6 +87,10 @@ pub static COMPATIBLE_ADDITIONAL_COLUMNS: LazyLock<HashMap<&'static str, HashSet
                     "collection_name",
                 ]),
             ),
+            (
+                MQTT_CONNECTOR,
+                HashSet::from(["topic", "offset", "partition"]),
```
[the remark this answers is not in the document]
Comment (Boudewijn26) [reply src/connector/src/parser/additional_columns.rs:92]: Yes, I followed the pattern established by the file connectors, which have "file" the same as the "partition" and read the `split_id` as the file

https://github.com/risingwavelabs/risingwave/blob/main/src/connector/src/parser/mod.rs#L484-L491

File: src/connector/src/parser/mod.rs [hunk 3]
```diff
@@ -494,6 +494,14 @@ impl SourceStreamChunkRowWriter<'_> {
                     // do special logic in `KvEvent::access_field`
                     parse_field(desc)
                 }
+                (_, &Some(AdditionalColumnType::Topic(_))) => {
+                    // topic is used as partition in mqtt connector
+                    return Ok(A::output_for(
+                        self.row_meta
+                            .as_ref()
+                            .map(|ele| ScalarRefImpl::Utf8(ele.split_id)),
```
[the remark this answers is not in the document]
Comment (Boudewijn26) [reply src/connector/src/parser/mod.rs:502]: The topic is saved in `split_id`, just like the "file" additional column is also stored in the `split_id`:
https://github.com/risingwavelabs/risingwave/blob/main/src/connector/src/source/mqtt/source/reader.rs#L96
https://github.com/risingwavelabs/risingwave/blob/main/src/connector/src/source/mqtt/source/message.rs#L41-L47
https://github.com/risingwavelabs/risingwave/blob/main/src/connector/src/source/mqtt/source/reader.rs#L97
https://github.com/risingwavelabs/risingwave/blob/main/src/connector/src/source/mqtt/source/message.rs#L29-L37

File: src/connector/src/source/mqtt/enumerator/mod.rs (hunk 1 above)
Comment (Boudewijn26) [reply src/connector/src/source/mqtt/enumerator/mod.rs:133]: I don't fully understand exactly what you're suggesting, but I'd say it's one of:
1. Move the logic inside the tokio thread to the `list_splits`, spawn the thread in `new`
2. During the call to `list_splits` spawn a thread and then stop that thread when the topics have come in
3. During the call to `list_splits` spawn a new thread to listen to the incoming mqtt topics
4. During the call to `list_splits` spawn a new thread if one isn't already running to listen to the incoming mqtt topics

Option 3 would spawn a new thread on each enumerator tick, which doesn't seem desirable to me
Option 4 is functionally identical to what we have just with added state to track the existence of the thread; I don't really see what would be gained
Option 2 would risk missing topics that come in after `list_splits` returns and come in before the next `tick`
Option 1 would require some inter thread communication in order to `poll` the mqtt eventloop inside the spawned thread whilst having the logic in the `list_splits`. That seems to make it more complicated than desired

Please indicate if I've misunderstood

File: src/connector/src/parser/additional_columns.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tabVersion) [reply src/connector/src/parser/additional_columns.rs:92]: Let's use `partition` instead. For other connectors, split_id is the partition number, no need for another additional column.

File: src/connector/src/source/mqtt/enumerator/mod.rs (hunk 1 above)
Comment (tabVersion) [reply src/connector/src/source/mqtt/enumerator/mod.rs:133]: > Option 2 would risk missing topics that come in after list_splits returns and come in before the next tick

Oh, I got it. Mqtt does not have persisted metadata so RisingWave needs a thread running in the background. The fix makes sense.

**Review by tabVersion — APPROVED** [review]
Please reuse `partition` for topic in MQTT connector. The concept is hard to align with other connectors. 

LGTM, thanks for your contribution.

**Comment by Boudewijn26** [comment]
@tabVersion With [`ae9f988` (#19017)](https://github.com/risingwavelabs/risingwave/pull/19017/commits/ae9f9881f6c9eb9063fea116aac304c5febb46b4) I've modified this so that it uses the `partition` instead of a separate `topic` column.

Meanwhile https://github.com/foundation-zero/risingwave-docs/commit/8d2b58fdf02bcffc5e36d5fbcad4c039aa686c73 is ready for a pull request once this is merged

**Review by yufansong — APPROVED** [review]
LGTM

**Review by xxhZs — APPROVED** [review]

**Comment by tabVersion** [comment]
Let's merge

**Pull request merged**